src/parallel/parallel.hpp

```diff
@@ -122,8 +122,11 @@ class SegmentLock {
     unlock();
 
 private:
-    std::mutex mutex_;
-    // Using () causes the linter to complain, thinking counter_ is a function.
+    // Use a recursive mutex, which allows multiple lock operations by the same
+    // thread on the lock. However, if the number of locks exceeds an
+    // unspecified limit, then it will throw a std::system_error.
+    std::recursive_mutex mutex_;
```

Comment:
I think I get why we're using a `recursive_mutex` but it still gives bad vibes. Is there any way we can do it with a regular mutex? Non-blocking though; we can revisit this later.

Reply:
I agree. Two ways of implementing it are as follows:

First, the thread manager could check what segment we're locking (and see if we've already locked it). This would be a user-space operation and could be in the lock manager. This provides better guarantees (e.g. no mysterious maximum number of locks). However, as currently implemented, it could be quite slow (linear lookup in a vector).

Otherwise, we could also expose the implementation details of the segment lock to the `insert` and `remove` methods. That way, they compute what segment they are locking. Since they know they lock in a monotonic direction, they only have to store the last segment locked and do at most one comparison. N.B. we could perform this optimization above, but it would be sketchy to abstract that detail away from the user.

Reply:
See #18 (wow, I created a circular reference)